### PR TITLE
Added support for multiple optional SCM's instances for GitLab & ADO

### PR DESCRIPTION
**Pull request by NimrodGolan**
### Description

> GitLab & Azure Devops are now supported with optional multiple SCM's instances

### References

> https://dev.azure.com/CxFlow/CxFlow/_workitems/edit/235
> https://dev.azure.com/CxFlow/CxFlow/_workitems/edit/236

### CODE REVIEW

**Review by AvivCX — COMMENTED** [review]
we need to add relevant page in our documentation

**Comment by NimrodGolan** [comment]
> we need to add relevant page in our documentation

We already have a documentation page: https://checkmarx.atlassian.net/wiki/spaces/IP/pages/2085389062/Multiple+SCM+s+Default+Instances

File: src/main/java/com/checkmarx/flow/config/RepoProperties.java [hunk 1]
```diff
@@ -47,6 +54,10 @@ public String getWebhookToken() {
         return webhookToken;
     }
 
+    public String getConfigWebhookToken(ControllerRequest controllerRequest) {
+        return scmConfigOverrider.determineConfigWebhookToken(this, controllerRequest);
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/config/RepoProperties.java:58]: Using the word "effective" in method names would be clearer, e.g. "getEffectiveWebhookToken".

File: src/main/java/com/checkmarx/flow/config/ScmConfigOverrider.java [hunk 2]
```diff
@@ -34,6 +34,7 @@ public String determineConfigWebhookToken(RepoProperties properties, ControllerR
     private String getScmOverriddenConfig(RepoProperties properties, ScmConfigParams configParams, String key) {
         String value;
         OptionalScmInstanceProperties optionalInstanceKey = properties.getOptionalInstances().get(key);
+
         if (optionalInstanceKey == null) {
             log.error("scm-instance key: {} does not exists on configuration file.", key);
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/config/ScmConfigOverrider.java:39]: "does not exist in"

File: src/main/java/com/checkmarx/flow/config/ScmConfigOverrider.java [hunk 3]
```diff
@@ -12,19 +12,19 @@
 @Slf4j
 public class ScmConfigOverrider {
 
-    public String determineConfigToken(RepoProperties properties, String scmInstance) {
+    String determineConfigToken(RepoProperties properties, String scmInstance) {
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/config/ScmConfigOverrider.java:15]: Can we replace the "determine..." methods with a single method that returns a DTO? The DTO will contain the result of merging instance properties and default properties.

File: src/main/java/com/checkmarx/flow/controller/ADOController.java [hunk 4]
```diff
@@ -239,8 +240,9 @@
             //build request object
             String gitUrl = repository.getRemoteUrl();
             log.debug("Using url: {}", gitUrl);
-            String gitAuthUrl = gitUrl.replace(HTTPS, HTTPS.concat(properties.getToken()).concat("@"));
-            gitAuthUrl = gitAuthUrl.replace(HTTP, HTTP.concat(properties.getToken()).concat("@"));
+            String configToken = properties.getConfigToken(controllerRequest.getScmInstance());
+            String gitAuthUrl = gitUrl.replace(HTTPS, HTTPS.concat(configToken).concat("@"));
+            gitAuthUrl = gitAuthUrl.replace(HTTP, HTTP.concat(configToken).concat("@"));
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/controller/ADOController.java:245]: The calls to "replace" can be chained.

File: src/main/java/com/checkmarx/flow/controller/GitLabController.java [hunk 5]
```diff
@@ -150,7 +148,13 @@ public String getTest() {
                     .filter(filter)
                     .build();
 
+            String mergeEndpoint = properties.getConfigApiUrl(request).concat(GitLabService.MERGE_NOTE_PATH);
+            mergeEndpoint = mergeEndpoint.replace("{id}", proj.getId().toString());
+            mergeEndpoint = mergeEndpoint.replace("{iid}", objectAttributes.getIid().toString());
+            request.setMergeNoteUri(mergeEndpoint);
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/controller/GitLabController.java:154]: It's better to extract this code into a method.

File: src/main/java/com/checkmarx/flow/controller/GitLabController.java [hunk 6]
```diff
@@ -272,7 +259,30 @@ public String getTest() {
...
+                }
+                if (StringUtils.isNotEmpty(c.getUrl()) && bugType.equals(BugTracker.Type.GITLABCOMMIT)) {
+                    commitEndpoint = properties.getConfigApiUrl(request).concat(GitLabService.COMMIT_PATH);
+                    commitEndpoint = commitEndpoint.replace("{id}", proj.getId().toString());
+                    commitEndpoint = commitEndpoint.replace("{sha}", c.getId());
+                }
+            }
+            if(StringUtils.isNotEmpty(body.getUserEmail())) {
+                emails.add(body.getUserEmail());
+            }
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/controller/GitLabController.java:278]: Can we extract a method from here?

File: src/main/java/com/checkmarx/flow/config/RepoProperties.java [hunk 7]
```diff
@@ -1,11 +1,18 @@
...
 import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.annotation.Autowired;
 
 import javax.annotation.PostConstruct;
 import java.util.Map;
 
 public class RepoProperties {
+
+    @Autowired
+    private ScmConfigOverrider scmConfigOverrider;
```
[the remark this answers is not in the document]
Comment (NimrodGolan) [reply src/main/java/com/checkmarx/flow/config/RepoProperties.java:14]: I moved all logic to be read directly from ScmConfigOverrider and not from the properties.

File: src/main/java/com/checkmarx/flow/config/ScmConfigOverrider.java (hunk 2 above)
Comment (NimrodGolan) [reply src/main/java/com/checkmarx/flow/config/ScmConfigOverrider.java:39]: fixed

File: src/main/java/com/checkmarx/flow/controller/GitLabController.java (hunk 5 above)
Comment (NimrodGolan) [reply src/main/java/com/checkmarx/flow/controller/GitLabController.java:154]: Done

File: src/main/java/com/checkmarx/flow/controller/GitLabController.java (hunk 6 above)
Comment (NimrodGolan) [reply src/main/java/com/checkmarx/flow/controller/GitLabController.java:278]: Done

File: src/main/java/com/checkmarx/flow/config/ScmConfigOverrider.java (hunk 3 above)
Comment (NimrodGolan) [reply src/main/java/com/checkmarx/flow/config/ScmConfigOverrider.java:15]: Done

File: src/main/java/com/checkmarx/flow/controller/ADOController.java (hunk 4 above)
Comment (NimrodGolan) [reply src/main/java/com/checkmarx/flow/controller/ADOController.java:245]: Done.

**Review by AvivCX — APPROVED** [review]

**Comment by jbrotsos** [comment]
> > we need to add relevant page in our documentation
> 
> We already have a documentation page: https://checkmarx.atlassian.net/wiki/spaces/IP/pages/2085389062/Multiple+SCM+s+Default+Instances

This documentation is not public.

**Comment by AvivCX** [comment]
Thanks @jbrotsos.
I added this page to the wiki: 
https://github.com/checkmarx-ltd/cx-flow/wiki/SCM-Multi-instance

**Pull request merged**